docs/development/design/embed-api.rst

```diff
@@ -0,0 +1,154 @@
+Embed API
+=========
+
+The embedded API allows to embed content from docs pages in other sites.
+For a while it has been as an *experimental* feature without public documentation or real applications,
+but recently it has been used widely (mainly because we created a Sphinx extension).
+
+Due to this we need to have more friendly to use API,
+and general and stable enough to support it for a long time.
+
+.. contents::
+   :local:
+   :depth: 3
+
+Current implementation
+----------------------
+
+The current implementation of the API is partially documented in :doc:`/guides/embedding-content`.
+Some characteristics/problems are:
+
+- There are three ways of querying the API, and some rely on Sphinx's concepts like ``doc``.
+- Doesn't cache responses or doesn't purge the cache on build.
+- Doesn't support MkDocs.
+- It returns all sections from the current page.
+- Lookups are slow (~500 ms).
+- IDs returned aren't well formed (like empty IDs `#`).
+- The content is always an array of one element.
+- The section can be an identifier or any other four variants or the title of the section.
+- It doesn't return valid HTML for definition lists (``dd`` tags without a ``dt`` tag).
+- The client doesn't know if the page requires extra JS or CSS in order to make it work or look nice.
+
+Improvements
+------------
+
+These improvements aren't breaking changes, so we can implement them in the old and new API.
+
+- Support for MkDocs.
+- Always return a valid/well formed HTML block.
```

Comment:
If this refers to the `dd`/`dt` issue, I'm not sure if this will be easily doable for all the cases. We may end up re-parsing the `dd` to convert that HTML tag to another one, but that will break the style based on `dd` tags. So, this will be tricky to do it right and make it work for all the themes --which is also related with the new `extras` field.

Reply:
We just need to extract only that element. For example

```html
<dl class="foo">
 .. more..
 <dt>foo</dt> <dd>bar</dd>
.. more...
</dl>
```

becomes

```html
<dl class="foo">
 <dt>foo</dt> <dd>bar</dd>
</dl>
```

Extracting content isn't hard to do here.

Reply:
Hrm... I think that has the same problem.

When parsing `citation`/`glossary` we don't need to return the title/definition (`dt`), we only want the description (`dd`). Example in code https://github.com/readthedocs/readthedocs.org/blob/eb682d8b8f480e96e7478d968aa12e52f4118063/readthedocs/embed/views.py#L329-L335

Following, your example, to make it valid it should be:

```html
<dl class="foo">
 <dt></dt>
 <dd>bar</dd>
</dl>
```

However, this won't probably render correctly inside the tooltip, since in that context is not a description list --it's not a list at all-- but just the definition of one of the term from that list.

In those cases, `citation`/`glossary` will be rendered in the same than any other definition list and I'm not sure we want that. We need to think how we will support this. Example:

### Glossary (only includes the definition)

![Screenshot_2021-03-30_11-26-04](https://user-images.githubusercontent.com/244656/112966996-0fe25d00-914b-11eb-9c8a-256f5cb6b3f4.png)


### Config value (includes the definition list completely repeating the term hovered)

![Screenshot_2021-03-30_11-23-06](https://user-images.githubusercontent.com/244656/112967106-2e485880-914b-11eb-97ed-ed1df07e4b39.png)

Reply:
I think we should repeat the title, it doesn't necessary match the text where the tooltip is present. Removing the title if it's the same as the text can be done in the client side.

Reply:
Glossary and Citation special cases can be managed from the frontend/client since we have all the data we need (note that we return the `.next()` HTML element). We will just need to parse it remove the content we don't want to display.

_However_, for the `dt` case (definition of a config, Python method/function, etc) it won't work because we need the `.parent()` of HTML element picked by ID. Take a look at this example:

```html
<dl class="std confval">
  <dt id="confval-hoverxref_role_types">
    <code class="sig-name descname">hoverxref_role_types</code>
    <a class="headerlink" href="#confval-hoverxref_role_types" title="Permalink to this definition">¶</a>
  </dt>
  <dd>
    <p>Description: Style to use by default when hover each type of reference (role).</p>
    <!-- ... snip ... -->
  </dd>
</dl>
```

(example from https://sphinx-hoverxref.readthedocs.io/en/latest/configuration.html#confval-hoverxref_role_types)

With the id `confval-hoverxref_role_types` we are currently returning the whole `dl` because we access to the _parent_ element. If we return just the HTML tag matched, we will be returning only the `dt` which does not include the description nor any content.

Without handling this special case properly, we will show tooltips like:

![Screenshot_2021-05-27_16-16-41](https://user-images.githubusercontent.com/244656/119842522-0a18b600-bf07-11eb-9b4d-31bcf21cd784.png)



We need to find a general solution for these special cases.

Reply:
@humitos not sure to understand the problem there? that's the same example we have in the document. The id is on the dt element, so we fetch the dt and the next dd enclosed in the dl tag.

Reply:
I may not have expressed myself properly.

This is a _special case_ because we are manipulating the HTML to manage a particular case on Sphinx: **return the parent of a particular ID, instead of itself**. Since we are going to support external sites that will not be generated with Sphinx (*), we can't assume that we always have to do this special HTML manipulation. That's why I'm saying that we need a general solution to differentiate these special cases from regular (non-Sphinx) pages.

I'm thinking that we will need to communicate this to the backend somehow. Maybe a request GET argument: `?doctool=sphinx&version=3.5.1`

(*) or even with a different version of Sphinx that may adds the `id=` in the `dl`.

Reply:
that's part of how we expect the html to be, following semantic html and using correct tags, if a site has malformed tags or if the id is in a different element we don't try to fix it, we just return it as is.